Wikipedia/Networking/Serializers/WMFMantleJSONResponseSerializer.m

```diff
@@ -0,0 +1,111 @@
+
+@end
+
+@interface WMFMantleJSONObjectResponseSerializer : WMFMantleJSONResponseSerializer
+
+@end
+
+@interface WMFMantleJSONCollectionResponseSerializer : WMFMantleJSONResponseSerializer
+
+@end
+
+@implementation WMFMantleJSONResponseSerializer
+
++ (instancetype)serializerForCollectionsOf:(Class __nonnull)model fromKeypath:(NSString* __nonnull)keypath {
+    return [[WMFMantleJSONCollectionResponseSerializer alloc] initWithModelClass:model jsonKeypath:keypath];
+}
+
++ (instancetype)serializerForInstancesOf:(Class __nonnull)model fromKeypath:(NSString* __nonnull)keypath {
+    return [[WMFMantleJSONObjectResponseSerializer alloc] initWithModelClass:model jsonKeypath:keypath];
+}
+
+- (instancetype)initWithModelClass:(Class __nonnull)modelClass jsonKeypath:(NSString* __nonnull)keypath {
+    self = [super init];
+    if (self) {
+        _modelClass  = modelClass;
+        _jsonKeypath = [keypath copy];
+    }
+    return self;
+}
+
+- (id)responseObjectForResponse:(NSURLResponse*)response
+                           data:(NSData*)data
+                          error:(NSError* __autoreleasing*)error {
+    NSDictionary* json = [super responseObjectForResponse:response data:data error:error];
+    if (!json) {
+        return nil;
+    }
+    id value = self.jsonKeypath.length ? [json valueForKeyPath : self.jsonKeypath] : json;
+    if (!value && self.jsonKeypath.length) {
+        DDLogWarn(@"No value returned when serializing %@ with keypath %@ from response: %@",
+                  self.modelClass, self.jsonKeypath, json);
+    }
+    return value;
+}
+
+- (NSString*)description {
+    return [NSString stringWithFormat:@"%@ %@ %@", [super description], self.modelClass, self.jsonKeypath];
+}
+
+@end
+
+@implementation WMFMantleJSONObjectResponseSerializer
+
+- (id)responseObjectForResponse:(NSURLResponse*)response
+                           data:(NSData*)data
+                          error:(NSError* __autoreleasing*)error {
+    NSDictionary* jsonObject = [super responseObjectForResponse:response data:data error:error];
+    if (!jsonObject) {
+        return nil;
+    } else if (![jsonObject isKindOfClass:[NSDictionary class]]) {
+        DDLogError(@"%@ expected dictionary value, got: %@", self, jsonObject);
+    }
+    return [MTLJSONAdapter modelOfClass:self.modelClass fromJSONDictionary:jsonObject error:error];
+}
+
+@end
+
+@implementation WMFMantleJSONCollectionResponseSerializer
+
+- (id)responseObjectForResponse:(NSURLResponse*)response
+                           data:(NSData*)data
+                          error:(NSError* __autoreleasing*)error {
+    id value = [super responseObjectForResponse:response data:data error:error];
+    if ([value isKindOfClass:[NSArray class]]) {
+        return [MTLJSONAdapter modelsOfClass:self.modelClass fromJSONArray:value error:error];
+    } else if (value) {
+        // most MW API responses are indexed by page ID, just grab all the values in the dictionary
```

Comment:
If we want to refactor the serializers to make them more reusable, we should be sure to make it configurable in the ways we need it to be. In other words, it should work without making assumptions about the data (explicit vs implicit).

In this case, it seems like we should be able to specify a mapping from a dictionary to an array. Something like this comes to mind (you probably have a better idea):

```
@property (nonatomic, copy) WMFCollectionTransformBlock transformBlock;
```